### DISCUSSION TITLE
Why is this method ineffective?

### DISCUSSION THREAD

**Discussion by peakedshout** [fyne-io/fyne#4607] [Q&A]
type Unit struct {
	wStart, wEnd float32
	hStart, hEnd float32
	fyne.CanvasObject
}

Like this, I made a combinatory, but it doesn't show up in use? This looks like the method should be implemented

**Comment by Jacalz** [comment]
It is just a data, it doesn't contain any logic. What are you trying to do? Extending `fyne.CanvasObject` basically doesn't do anything because it is just an interface, not a concrete implementation.

**Reply by peakedshout** [reply]
what?I guess I didn't express it clearly enough, like the following test code:
```
func tmp() {
	a := app.New()
	w := a.NewWindow("test")

	type unit struct {
		fyne.CanvasObject
	}
	box := container.NewVBox(
		&unit{CanvasObject: widget.NewLabel("????")},
		widget.NewLabel("????"),
	)
	w.Resize(fyne.NewSize(100, 100))
	w.SetContent(box)
	w.ShowAndRun()
}
```
In the box, the components of the combined unit are not displayed, and this is what puzzles me. Or is fyne not called internally based on the interface implementation?

**Reply by Jacalz** [reply]
You can extend widgets, not canvas objects. See https://docs.fyne.io/extend/extending-widgets. This one might also be useful: https://docs.fyne.io/extend/numerical-entry.

Passing in the label to the unit type doesn't make it call any of the methods of the label when invoked. Writing a type like your unit type by extending an interface just makes the compiler generate empty methods for everything in the interface.

What are you trying to accomplish with that type?

**Reply by peakedshout** [reply]
I want to implement a display aspect ratio for any object, so that the layout can be resized according to these proportions, and not limited to a certain component object.

**Reply by andydotxyz** [reply]
You should look at layouts for this functionality, not extending the generic concept of a canvas object.
It may seem counter-intuitive at first, but the position and size of widgets is controlled by the layout of the container they are in.
With this method we can adapt UI much more easily to the variety of output devices we support.
See https://docs.fyne.io/extend/custom-layout

**Reply by peakedshout** [reply]
I did implement a custom layout that I expected to be displayed according to the scale, so that the properties of each scale should not be bound to the layout object?

**Reply by peakedshout** [reply]
In the custom layout, I get the extended layout object, and the layout will be resized according to the extended properties of the layout object, I don't think this idea should be too much of a problem?

**Reply by andydotxyz** [reply]
I don't understand what an extended layout object is... A layout is any type that implements the two methods of the `Layout` interface. No extension and no extra objects should be needed.

**Comment by andydotxyz** [comment]
> I did implement a custom layout that I expected to be displayed according to the scale, so that the properties of each scale should not be bound to the layout object?

I don't fully understand this, but perhaps you can see how `Border` container uses a layout with extra properties to influence the layout. Or check out the responsive layout in fyne-x repo https://github.com/fyne-io/fyne-x/blob/master/layout/responsive.go for inspiration.

**Reply by peakedshout** [reply]
```
func (l *Layout) Layout(objects []fyne.CanvasObject, size fyne.Size) {
	for _, one := range objects {
		if !one.Visible() {
			continue
		}
		unit := one.(*LayoutObj)
		var ws, we, hs, he float32
		if l.proportional {
			ws = size.Width * (unit.WStart)
			we = size.Width * (unit.WEnd)
			hs = size.Height * (unit.HStart)
			he = size.Height * (unit.HEnd)
		} else {
			ws = unit.WStart
			we = unit.WEnd
			hs = unit.HStart
			he = unit.HEnd
		}
		unit.Resize(fyne.NewSize(we-ws, he-hs))
		one.Move(fyne.NewPos(ws, hs))
	}
}
```
I think the above piece of code should be able to express what I want to achieve

**Reply by andydotxyz** [reply]
Not really. You seem to be asking the widgets where they would like to appear - that's not really what layouts do.
I may be missing something - but can't your layout make the decision based on available space?

**Reply by peakedshout** [reply]
I just want to show it proportionally based on the total size of the current layout.
Perhaps, I could extend a widget to wrap this unidentified object, and expand the information in this new component. Alternatively, append a map to the layout to do so.
However, in the case of golang, it is not uncommon to extend the approach of the interface (language features). As defined by the API, it accepts the CanvasObject interface as a parameter, and the external implementation of the interface does not play a role, which is undoubtedly very confusing. For a golang programmer who is new to fyne, there is a high probability that this trap will be triggered, especially those who like to design abstractly with interfaces.

**Reply by andydotxyz** [reply]
As Jacob mentioned you can provide custom widgets - but not custom canvas objects, because the painter will not understand how to paint them.

Going back to the original point the widgets should not control layout - it's not how Fyne is designed so you will be constantly fighting the API features.
Consider instead using the responsive layout from fyne-x https://github.com/fyne-io/fyne-x

**Reply by peakedshout** [reply]
Well, I don't want to argue about anything anymore, I can only hope that Fyne will get better and have fewer pitfalls like that.

**Reply by andydotxyz** [reply]
I'm closing out this discussion - but I hope you can see that in the answers above we have shown how to do each of the things you desired, and explained why the API is designed the way it is.
